pkg/policycache/policy_map.go

```diff
@@ -112,91 +86,56 @@ func (m *pMap) update(old kyverno.PolicyInterface, new kyverno.PolicyInterface)
+				Generate:             sets.NewString(),
+				VerifyImagesMutate:   sets.NewString(),
+				VerifyImagesValidate: sets.NewString(),
 			}
-
+		}
+		if rule.HasMutate() {
+			m.kindDataMap[kind][Mutate] = m.kindDataMap[kind][Mutate].Insert(pName)
 			continue
 		}
-
 		if rule.HasValidate() {
 			if enforcePolicy {
-				if !m.nameCacheMap[ValidateEnforce][kind+"/"+pName] {
-					m.nameCacheMap[ValidateEnforce][kind+"/"+pName] = true
-					validatePolicy := m.kindDataMap[kind][ValidateEnforce]
-					m.kindDataMap[kind][ValidateEnforce] = append(validatePolicy, pName)
-				}
-
-				continue
+				m.kindDataMap[kind][ValidateEnforce] = m.kindDataMap[kind][ValidateEnforce].Insert(pName)
+			} else {
+				m.kindDataMap[kind][ValidateAudit] = m.kindDataMap[kind][ValidateAudit].Insert(pName)
 			}
-
-			if !m.nameCacheMap[ValidateAudit][kind+"/"+pName] {
-				m.nameCacheMap[ValidateAudit][kind+"/"+pName] = true
-				validatePolicy := m.kindDataMap[kind][ValidateAudit]
-				m.kindDataMap[kind][ValidateAudit] = append(validatePolicy, pName)
-			}
-
 			continue
 		}
-
 		if rule.HasGenerate() {
-			if !m.nameCacheMap[Generate][kind+"/"+pName] {
-				m.nameCacheMap[Generate][kind+"/"+pName] = true
-				generatePolicy := m.kindDataMap[kind][Generate]
-				m.kindDataMap[kind][Generate] = append(generatePolicy, pName)
-			}
-
+			m.kindDataMap[kind][Generate] = m.kindDataMap[kind][Generate].Insert(pName)
 			continue
 		}
-
 		if rule.HasVerifyImages() {
-			if !m.nameCacheMap[VerifyImagesMutate][kind+"/"+pName] {
-				m.nameCacheMap[VerifyImagesMutate][kind+"/"+pName] = true
-				imageVerifyMapPolicy := m.kindDataMap[kind][VerifyImagesMutate]
-				m.kindDataMap[kind][VerifyImagesMutate] = append(imageVerifyMapPolicy, pName)
-			}
-
+			m.kindDataMap[kind][VerifyImagesMutate] = m.kindDataMap[kind][VerifyImagesMutate].Insert(pName)
 			if rule.HasImagesValidationChecks() {
-				m.nameCacheMap[VerifyImagesValidate][kind+"/"+pName] = true
-				imageVerifyMapPolicy := m.kindDataMap[kind][VerifyImagesValidate]
-				m.kindDataMap[kind][VerifyImagesValidate] = append(imageVerifyMapPolicy, pName)
+				m.kindDataMap[kind][VerifyImagesValidate] = m.kindDataMap[kind][VerifyImagesValidate].Insert(pName)
 			}
-
 			continue
 		}
 	}
 }
 
-func removeCacheHelper(rmr kyverno.ResourceFilter, m *pMap, pName string) {
-	for _, gvk := range rmr.Kinds {
-		_, kind := kubeutils.GetKindFromGVK(gvk)
-		dataMap := m.kindDataMap[kind]
-		for policyType, policies := range dataMap {
-			var newPolicies []string
-			for _, p := range policies {
-				if p == pName {
-					continue
-				}
-				newPolicies = append(newPolicies, p)
-			}
-			m.kindDataMap[kind][policyType] = newPolicies
-		}
-		for _, nameCache := range m.nameCacheMap {
-			if ok := nameCache[kind+"/"+pName]; ok {
-				delete(nameCache, kind+"/"+pName)
-			}
+func removeCacheHelper(match kyverno.MatchResources, m *pMap, pName string) {
+	for _, gvk := range match.GetKinds() {
+		kind := getKind(gvk)
```

Comment:
Looks like we were missing a call to `kubeutils.SplitSubresource` here.
@realshuting shall we back port to 1.7 ?